crypto/cms/cms_lib.c

```diff
@@ -20,18 +20,65 @@ DEFINE_STACK_OF(CMS_RevocationInfoChoice)
 DEFINE_STACK_OF(X509)
 DEFINE_STACK_OF(X509_CRL)
 
-IMPLEMENT_ASN1_FUNCTIONS(CMS_ContentInfo)
 IMPLEMENT_ASN1_PRINT_FUNCTION(CMS_ContentInfo)
+IMPLEMENT_ASN1_ENCODE_FUNCTIONS_fname(CMS_ContentInfo, CMS_ContentInfo, CMS_ContentInfo)
+
+
+static void cms_set0_libctx(CMS_ContentInfo *cms, OPENSSL_CTX *libctx,
+                            const char *propq)
+{
+    if (cms != NULL) {
+        cms->ctx.libctx = libctx;
+        cms->ctx.propq = propq;
+    }
+}
+
+CMS_ContentInfo *CMS_ContentInfo_new(void)
+{
+    CMS_ContentInfo *ci;
+
+    ci = (CMS_ContentInfo *)ASN1_item_new(ASN1_ITEM_rptr(CMS_ContentInfo));
+    cms_set0_libctx(ci, NULL, NULL);
+    return ci;
+}
+
+CMS_ContentInfo *CMS_ContentInfo_new_ex(OPENSSL_CTX *libctx, const char *propq)
+{
+    CMS_ContentInfo *ci = CMS_ContentInfo_new();
+
+    cms_set0_libctx(ci, libctx, propq);
+    return ci;
+}
+
+void CMS_ContentInfo_free(CMS_ContentInfo *cms)
+{
+    ASN1_item_free((ASN1_VALUE *)cms, ASN1_ITEM_rptr(CMS_ContentInfo));
+}
+
+const CMS_CTX *CMS_get0_ctx(const CMS_ContentInfo *cms)
+{
+    return cms != NULL ? &cms->ctx : NULL;
+}
+
+void CMS_set0_libctx(CMS_ContentInfo *ci, OPENSSL_CTX *libctx,
```

Comment:
Should this be `CMS_ContentInfo_set0_libctx`? Same for the get0 above

Reply:
No - the CMS_ is used all through tthe APIS's to mean contentinfo..